django/test/utils.py

```diff
@@ -640,3 +642,47 @@ def setUp(self):
 
     def tearDown(self):
         self.logger.handlers[0].stream = self.old_stream
+
+
+class isolated_apps(object):
+    def __init__(self, *installed_apps):
+        self.installed_apps = installed_apps
+
+    def enable(self):
+        self.old_apps = Options.default_apps
+        apps = Apps(self.installed_apps)
+        setattr(Options, 'default_apps', apps)
+        return apps
+
+    def disable(self):
+        setattr(Options, 'default_apps', self.old_apps)
+
+    def __enter__(self):
+        return self.enable()
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.disable()
+
+    def __call__(self, decorated):
+        if isinstance(decorated, type):
+            # A class is decorated
+            decorated_setUp = decorated.setUp
+            decorated_tearDown = decorated.tearDown
+
+            def setUp(inner_self):
+                inner_self.apps = self.enable()
+                decorated_setUp(inner_self)
+
+            def tearDown(inner_self):
+                decorated_tearDown(inner_self)
+                self.disable()
+
+            decorated.setUp = setUp
+            decorated.tearDown = tearDown
+            return decorated
+        else:
+            @wraps(decorated)
+            def inner(*args, **kwargs):
+                with self:
+                    return decorated(*args, **kwargs)
```

Comment:
I was hesitant to pass the isolated registry as an argument here since this is not a pattern used by other test decorators.